src/libraries/Microsoft.PowerFx.Core/Public/Values/FormulaValue.cs

```diff
@@ -366,57 +291,74 @@ internal static RecordValue RecordFromJsonObject(JsonElement element)
 
         #region Host Tables API
 
-        /// <summary>
-        /// Create a table from an untyped IEnumerable. This can be useful in some dynamic scenarios.
-        /// </summary>
-        /// <param name="values"></param>
-        /// <param name="elementType"></param>
-        /// <returns></returns>
-        internal static TableValue TableFromEnumerable(
-            System.Collections.IEnumerable values,
-            Type elementType)
+        public static TableValue NewTable<T>(TypeMarshallerCache cache, params T[] array)
+        {
+            return NewTable(cache, (IEnumerable<T>)array);
+        }
+
+        // If T is a RecordValue or FormulaValue, this will call out to that overload. 
+        public static TableValue NewTable<T>(TypeMarshallerCache cache, IEnumerable<T> rows)
         {
-            if (elementType == null)
+            if (cache == null)
+            {
+                throw new ArgumentNullException(nameof(cache));
+            }
+
+            // For FormulaValue, we need to inspect the runtime type. 
+            if (rows is IEnumerable<RecordValue> records)
+            {
+                return NewTable(records);
+            }
+
+            // Check FormulaValue after RecordValue
+            if (rows is IEnumerable<FormulaValue> values)
             {
-                throw new ArgumentNullException(nameof(elementType));
+                // Check for a single-column table.                
+                return NewSingleColumnTable(values);
             }
 
-            var records = TableToRecords(values, elementType);
-            var recordType = (RecordType)records.First().IRContext.ResultType;
-            return TableFromRecords(records, recordType.ToTable());
+            // Statically marshal the T to a FormulaType.             
+            var value = cache.Marshal(rows);
+            return (TableValue)value;
         }
 
-        public static TableValue TableFromRecords(params RecordValue[] values)
+        /// <summary>
+        /// Construct a table from records. Assumed that Records must be the same type. 
```

Comment:
>Assumed that Records must be the same type. [](http://example.com/codeflow?start=44&length=43)

Should we union the types? For example, Table({a:1}, {b:2}, {a:3}) creates a result of type *[a:n,b:n]

Reply:
I think we should just require the type to be passed in and be explicit.  This is getting into the kind of analysis the binder does. 
We can add a future overload that tries to be smarter.